net: openthread: add pskd join loop

Philipp-Wohlgenannt-Tridonic-com opened the pull request:
restart joiner if in autostart mode and join hasn't succeded yet or if joiner is back to idle

Signed-off-by: Philipp Wohlgenannt <philipp.wohlgenannt@tridonic.com>

Philipp-Wohlgenannt-Tridonic-com commented:
i will fix glint and checkpath and then i will commit it again

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -246,13 +259,55 @@ static void openthread_process(void *context, void *arg2, void *arg3)
 {
 	struct openthread_context *ot_context = context;
 
+		// not restarting the joiner before a changing delay has passed
```
We do no use `//` comments in zephyr, could you change these to `/* */`

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -246,13 +259,55 @@ static void openthread_process(void *context, void *arg2, void *arg3)
 	struct openthread_context *ot_context = context;
 
+		// not restarting the joiner before a changing delay has passed
+		#define JOINER_DELAY_INITIAL 2000
```
Usually we put these defines somewhere in the start of the file.

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -246,13 +259,55 @@ static void openthread_process(void *context, void *arg2, void *arg3)
+		#define JOINER_DELAY_MAX 32000
+		// if no join in 10 minutes, reboot with better luck
+		#define JOINER_REBOOT_TIME K_MINUTES(10)
+		s64_t joiner_delay_ms = JOINER_DELAY_INITIAL;
```
Indentation is wrong for these changes

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -246,13 +259,55 @@ static void openthread_process(void *context, void *arg2, void *arg3)
+			if (k_uptime_get() - reboot_ref_ms >
+				JOINER_REBOOT_TIME) {
+				// printk doesn't seem to do the job
+				printf("Joining failed for %d seconds..."
```
Please use logging macros

Philipp-Wohlgenannt-Tridonic-com replied:
Logging will not show up on UART, as the system is rebooting immediately

jukkar replied:
Ok, please change this to `printk()` as that is using less resources.

Philipp-Wohlgenannt-Tridonic-com replied:
Printk doesn't work in this case. A explanation fr that might be in the discussion in https://github.com/zephyrproject-rtos/zephyr/pull/25627

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -246,13 +259,55 @@ static void openthread_process(void *context, void *arg2, void *arg3)
+				sys_reboot(SYS_REBOOT_COLD);
+			}
+			//delay before restarting joiner
+			k_sleep(joiner_delay_ms);
```
If the value is in ms, then use `k_msleep()` instead

rlubos replied:
We definitely should not block the OpenThread processing thread. In general, I don't think we should keep this logic in the main loop.

OpenThread calls `ot_joiner_start_handler` function when joining procedure is complete (successful or not). In case we need to retry the procedure, we should keep the retry logic in there. If the retry attempt has to be delayed, we should schedule a delayed work.

And in general, the retry-parameters should be configurable (i.e. initial delay, consider replacing max delay and reboot time with maximum retry count).

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -246,13 +259,55 @@ static void openthread_process(void *context, void *arg2, void *arg3)
+			if (joiner_delay_ms > JOINER_DELAY_MAX) {
+				joiner_delay_ms = JOINER_DELAY_INITIAL;
+			}
+			NET_WARN("Restarting joiner.");
```
Please place empty line after } that is ending a block

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -246,13 +259,55 @@ static void openthread_process(void *context, void *arg2, void *arg3)
+			}
+			//delay before restarting joiner
+			k_sleep(joiner_delay_ms);
+			joiner_delay_ms *= 2;
```
Probably compiler will optimize this but we could write this as `joiner_delay_ms <<= 1`

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -246,13 +259,55 @@ static void openthread_process(void *context, void *arg2, void *arg3)
+				NET_ERR("Failed to start joiner [%d]", error);
+			}
+		}
 		k_sem_take(&ot_sem, K_FOREVER);
```
here too, empty line after ending }

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -360,8 +415,11 @@ static void openthread_start(struct openthread_context *ot_context)
+		(void) devicedata_get_pskd(&pPSKD);
+		joiner_autostart_started = true;
+		error = otJoinerStart(ot_instance, pPSKD, NULL,
+		// error = otJoinerStart(ot_instance, OT_JOINER_PSKD, NULL,
```
Just remove code that is not used

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -96,6 +96,18 @@ static struct net_linkaddr *ll_addr;
 
+#include <power/reboot.h>
+#include <stdio.h>
+int devicedata_get_pskd(const char **);
```
These additions look very "tight", we could put empty line after the includes, and before the devicedata_get_pskd()

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -96,6 +96,18 @@ static struct net_linkaddr *ll_addr;
+#include <power/reboot.h>
+#include <stdio.h>
+int devicedata_get_pskd(const char **);
+const char *pPSKD = OT_JOINER_PSKD;
```
Shouldn't this be static?

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -96,6 +96,18 @@ static struct net_linkaddr *ll_addr;
+const char *pPSKD = OT_JOINER_PSKD;
+static bool joiner_autostart_started;
+static bool joiner_successful;
+__weak int devicedata_get_pskd(const char **pskd)
```
Why this is __weak?

Philipp-Wohlgenannt-Tridonic-com replied:
so it gives a default implementation, which can be easily overridden by application code

jukkar replied:
Sure, but why application would need or want to override this?

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -360,8 +415,11 @@ static void openthread_start(struct openthread_context *ot_context)
-
-		error = otJoinerStart(ot_instance, OT_JOINER_PSKD, NULL,
+		// should this fail, we'll stick to OT_JOINER_PSKD
+		(void) devicedata_get_pskd(&pPSKD);
```
If the result is ignored anyway, I think the function should not return the result in a first place (i. e. be of `void` type).

Philipp-Wohlgenannt-Tridonic-com replied:
maybe the return value of the app-defined function could be used elsewhere, but it is not interesting here?

rlubos reviewed (changes requested):
Some high-level comments for now.

Philipp-Wohlgenannt-Tridonic-com commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -246,13 +259,55 @@ static void openthread_process(void *context, void *arg2, void *arg3)
+			== OT_JOINER_STATE_IDLE) {
+			if (k_uptime_get() - reboot_ref_ms >
+				JOINER_REBOOT_TIME) {
+				// printk doesn't seem to do the job
```
this line can be removed

Philipp-Wohlgenannt-Tridonic-com commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -96,6 +96,18 @@ static struct net_linkaddr *ll_addr;
 
+#include <power/reboot.h>
+#include <stdio.h>
+int devicedata_get_pskd(const char **);
```
@rlubos: will you make that openthread API or should we make a suggestion how it could look like?

we are reworking the rest of the code so it will take a bit till i will push it again. Thanks for your help!

rlubos replied:
I already suggested the rename: `openthread_get_psdd`. By making it available in the API I just meant that we should put the declaration in `include/net/openthread.h`, a public header that is available for applications. Users need to know what function they should implement.

Philipp-Wohlgenannt-Tridonic-com replied:
I'm actually confused... why does it need to be declared in openthread.h in any case? I see it if it was openthread_SET_pskd(), which would be called by app... but currently it works the other way around

rlubos replied:
Yes, it works the other way, but still, it needs to be documented how the function should look like. How is the user supposed to know what function should he implemented and how should it look like, if we don't provide the expected function declaration?

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -361,7 +431,13 @@ static void openthread_start(struct openthread_context *ot_context)
-		error = otJoinerStart(ot_instance, OT_JOINER_PSKD, NULL,
+		/* should this fail, we'll stick to OT_JOINER_PSKD */
+		openthread_get_pskd(&joiner_pskd);
+		u32_t join_delay = sys_rand32_get();
```
What's with those delays? We can't hold up the entire system for up to 15 seconds. The commit message does not mention it at all.

Philipp-Wohlgenannt-Tridonic-com replied:
I accidentally pushed that part too but that should not be in here. There is a different open pull request for this so i will remove it here.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -401,6 +477,13 @@ static int openthread_init(struct net_if *iface)
 	NET_DBG("openthread_init");
 
+	if (!(IS_ENABLED(CONFIG_OPENTHREAD_NCP))) {
+		u32_t boot_delay;
```
Same story here.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -361,7 +431,13 @@ static void openthread_start(struct openthread_context *ot_context)
 		NET_DBG("Starting OpenThread join procedure.");
 
-		error = otJoinerStart(ot_instance, OT_JOINER_PSKD, NULL,
+		/* should this fail, we'll stick to OT_JOINER_PSKD */
```
This comment seems no longer valid.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -227,17 +239,75 @@ void ot_receive_handler(otMessage *aMessage, void *context)
+
+static void ot_joiner_restart_timer(void *dummy)
+{
+	k_work_submit(&joiner_restart);
```
You could simply use `k_delayed_work_submit` instead of timer + work combo.

Philipp-Wohlgenannt-Tridonic-com replied:
We have tried that but that did not work with our zephyr devices.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -227,17 +239,75 @@ void ot_receive_handler(otMessage *aMessage, void *context)
+		if (IS_ENABLED(CONFIG_OPENTHREAD_JOINER_CONTINUOUS)) {
+			join_attempts++;
+
+			if (CONFIG_OPENTHREAD_JOINER_REBOOT != 0) {
```
It would be more concise to join the two `if` cases and use `&&` operator.

Philipp-Wohlgenannt-Tridonic-com replied:
It was meant that with CONFIG_OPENTHREAD_JOINER_CONTINUOUS we can activate the joiner loop. In the joiner loop there are 3 more configurable options:

1. - CONFIG_OPENTHREAD_JOINER_ADD_DELAY_MS: Added joining delay [ms] till next try.
2. - CONFIG_OPENTHREAD_JOINER_TRIALS_MAX_DELAY: Number of joining trials till the maximum delay is reached. If it is configured with 0, there is no maxmium delay.
3. - CONFIG_OPENTHREAD_JOINER_REBOOT: Reboots the device after a specified number of failed joining trials. If it is configured with 0, it never reboots.

That's why I made a few if statements in another if statement. With CONFIG_OPENTHREAD_JOINER_CONTINUOUS && CONFIG_OPENTHREAD_JOINER_REBOOT  these two would basically be the same option and there is no possibility to start the joining loop in an endless loop.

rlubos replied:
I understand the overall structure of this joiner loop, my comment only referred to the line we're commenting on, i.e. to have:
```
if ((CONFIG_OPENTHREAD_JOINER_REBOOT != 0) &&
    (join_attempts >= CONFIG_OPENTHREAD_JOINER_REBOOT)) {
```
but it's not a big deal really, just one less indentation level - less line wrapping.

Philipp-Wohlgenannt-Tridonic-com replied:
Sorry my mistake I  looked at if (IS_ENABLED(CONFIG_OPENTHREAD_JOINER_CONTINUOUS) as second if. Of course I will make one if out of these 2.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -227,17 +239,75 @@ void ot_receive_handler(otMessage *aMessage, void *context)
+			if (CONFIG_OPENTHREAD_JOINER_REBOOT != 0) {
+				if (join_attempts >=
+					CONFIG_OPENTHREAD_JOINER_REBOOT) {
+					printf("Joining failed %d times ... "
```
Please, let's not introduce a bulky `printf` to the final binary. We should use logger here, just as in other places. If you're not getting an output because of the reboot, there's a function for that:
https://docs.zephyrproject.org/latest/reference/logging/index.html#_CPPv49log_panicv

rlubos reviewed (commented):
Thanks for the rework, the retransmit logic looks much better now. There are still some issues left though.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -227,17 +240,74 @@ void ot_receive_handler(otMessage *aMessage, void *context)
 {
 	struct openthread_context *ot_context = context;
-
+	static u32_t restart_delay;
```
Also, looking at PR #25679 which got just merged,it seems we should use `uint32_t` now.

Philipp-Wohlgenannt-Tridonic-com commented:
We have reverted it to to timer implementation because the previous versions with k_work_submit() or k_delayed_work_submit() caused crashes on our local zephyr devices.

pasi-ojala-tridonic commented:
> We have reverted it to to timer implementation because the previous versions with k_work_submit() or k_delayed_work_submit() caused crashes on our local zephyr devices.

Just as a clarifying comment: it was the use k_delayed_work item that was causing us (so far non-understood) crashes in certain use cases of the joiner, so we went back to using the timer and k_work, which survives the same case fine.

rlubos reviewed (approved):
Just please address the Shippable failure (currently this code does not build due to timer handler function signature mismatch).

Philipp-Wohlgenannt-Tridonic-com committed:
net: openthread: add pskd join loop

restart joiner if:
-in autostart mode and join hasn't succeeded yet
-joiner is back to idle

Signed-off-by: Philipp Wohlgenannt <philipp.wohlgenannt@tridonic.com>
